packages/dev/docs/pages/react-spectrum/forms.mdx

```diff
@@ -0,0 +1,449 @@
+</Form>
+```
+
+### Realtime validation
+
+Usually, validation errors should be displayed to the user after the value is committed (e.g. on blur), or when the form is submitted. This avoids confusing the user with irrelevant errors while they are still entering a value.
+
+In some cases, validating in realtime can be desireable, such as when meeting password requirements. This can be accomplished by making the field value [controlled](#controlled-forms), and setting the `isInvalid` and `errorMessage` props accordingly.
+
+```tsx example
+function Example() {
+  let [password, setPassword] = React.useState('');
+  let error;
+  if (password.length < 8) {
+    error = 'Password must be 8 characters or more.';
+  } else if ((password.match(/[A-Z]/g) ?? []).length < 2) {
+    error = 'Password must include at least 2 upper case letters';
+  } else if ((password.match(/[^a-z]/ig) ?? []).length < 2) {
+    error = 'Password must include at least 2 symbols.';
+  }
+
+  return (
+    <TextField
+      label="Password"
+      /*- begin highlight -*/
+      isInvalid={!!error}
+      errorMessage={error}
+      /*- end highlight -*/
+      value={password}
+      onChange={setPassword} />
+  );
+}
+```
+
+By default, the `isInvalid` prop does not block forms from being submitted. To block form submission, use the `validationBehavior="native"` prop on the [Form](Form.html) component.
+
+### Server validation
+
+Client side validation is useful to give the user immediate feedback, but only one half of the validation story. Data should also be validated on the backend for security and reliability, and your business logic may include rules which cannot be validated on the frontend.
+
+React Spectrum supports displaying server validation errors by passing the `validationErrors` prop to the [Form](Form.html) component. This should be set to an object that maps each field's `name` prop to a string or array of strings representing one or more errors. These are displayed to the user as soon as the `validationErrors` prop is set, and cleared after the user modifies each field's value.
+
+```tsx example
+function Example() {
+  /*- begin highlight -*/
+  let [errors, setErrors] = React.useState({});
+  /*- end highlight -*/
+  let onSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    let data = Object.fromEntries(new FormData(e.currentTarget));
+    let result = await callServer(data)
+    /*- begin highlight -*/
+    setErrors(result.errors);
+    /*- end highlight -*/
+  };
+
+  return (
+    <Form
+      validationBehavior="native"
+      maxWidth="size-3000"
+      onSubmit={onSubmit}
+      /*- begin highlight -*/
+      validationErrors={errors}
+      /*- end highlight -*/
+    >
+      <TextField label="Username" name="username" isRequired />
+      <TextField label="Password" name="password" type="password" isRequired />
+      <ButtonGroup>
+        <Button type="submit" variant="primary">Submit</Button>
```

Comment:
I can do the thing with this example where after correcting a field, it takes two clicks on the submit button to first clear the state then actually submit again

but it only takes one "Enter" key

Reply:
I think it's just because the page button under your mouse when the field blurs and the error message goes away, so when you mouse up you aren't over the button anymore.

Reply:
ahhh, yeah, that'd do it
that's unfortunate